Lib/test/test_threaded_pickle.py

```diff
@@ -0,0 +1,53 @@
+import sys
+import importlib
+import os
+import shutil
+import threading
+import unittest
+from test.support import TESTFN, forget
+import pickle
+
+slow_import_module = """
+import time
+
+time.sleep(%(delay)s)
```

Comment:
This test relies on sleep to work. I do not feel comfortable with such tests as it may fail or not work as intended almost certainly in the slowest or more exotic buildbots that we have. Check out, for example, the discussion in https://github.com/python/cpython/pull/10413 .

As an additional note, it also needs the `reap_threads` decorator.

Reply:
It will make the test a bit more complicated, but I think you can avoid `sleep` and make this more deterministic by creating a deliberate deadlock like this:

1. Acquire Lock A in the test
2. Attempt to acquire lock A in the partially-initialized module
3. Spawn a thread that attempts to unpickle the object that involves an import
4. Release lock A
5. After the partially-initialized module has lock A, it adds the attribute that the pickled object uses.

I think there's still a race condition there that doesn't make it deterministic - namely that the thread which is unpickling the object needs to reach the "access the imported attribute" stage *before* lock A is released in order for it to be a good test, but I think if you fiddle around with the locks and threads a little bit you can get it to work.

Reply:
I'll be more radical and say I don't think we need a test at all, if we can't make it simpler.

Reply:
@pitrou I think removing the test because it's not guaranteed to test the thing it's intended to test would be a mistake. This still successfully tests what it's intending to test with high probability, which means that a regression would be very unlikely to ever hit master. Unlike other tests which cause erroneous *failures*, this causes erroneous *successes*, so its flakiness is unlikely to cause any erroneously broken buildbots.

If left in as a "flaky" test, I'd probably view it similar to a `hypothesis` test, which is that it's a probabilistic test that never has false negatives - if the test succeeds you are only partially certain that you have no bugs, but if it fails you *know* you have a bug.

Elsewhere you mention adding a comment about why `PyImport_GetModule` was dropped to prevent regressions - if it's worth a comment (which is not guaranteed to work), it's probably worth at least *trying* to enforce this with a test.

Reply:
Well, there are corner cases that we fix without writing tests for, because they are too tricky to reproduce. I think this one is a bit borderline: it's possible to reproduce, but nevertheless it needs a complex machinery.

Reply:
I certainly agree that if it's likely to have a negative impact on the test suite for reasons *other* than that it's not guaranteed to fail in the case of a regression the test should probably be considered optional.

I do think it's worth trying to get a test for this in the test suite, even if just as an exercising in providing a good example of how to test a race condition.